internal/model.go

```diff
@@ -0,0 +1,229 @@
+	DeliveryModeAtLeastOnce   = "at_least_once"    // duplicates allowed, message loss not allowed
+	DeliveryModeExactlyOnce   = "exactly_once"     // most precise delivery mode
+	DeliveryModeMostOfTheTime = "most_of_the_time" // realistic delivery mode
+	DeliveryModeNeverEver     = "never_ever"       // in case of message fatigue
+
+	// event encodings
+
+	EncodingJson = "json"
+)
+
+type (
+	// A FilterChain is a slice of filter plugins
+	FilterChain struct {
+		Filters []*FilterPlugin `json:"filters"` // optional list of filter plugins that will be applied in order to perform arbitrary filtering and transformation functions
+	}
+
+	// A RoutingTable is a slice of routing entries and reprrsents the EARS routing table
+	RoutingTable []*Route
+
+	// A RoutingTableIndex is a hashmap mapping a routing entry hash to a routing entry pointer
+	RoutingTableIndex map[string]*Route
+
+	// A PluginIndex is a hashmap mapping a plugin instance hash to a plugin instance
+	PluginIndex map[string]Pluginer
+)
+
+type (
+	Pluginer interface {
+		Hasher
+		Validater
+		Initializer
+		Closer
+		SyncDoer
+		AsyncDoer
+		GetConfig() *Plugin
+		GetInputChannel() chan *Event
+		GetOutputChannel() chan *Event
+		GetRouteCount() int
+		GetEventCount() int
+	}
+
+	// A Hasher can provide a unique deterministic hash string based on its configuration parameters
+	Closer interface {
+		Close(ctx context.Context)
+	}
+
+	// A Hasher can provide a unique deterministic hash string based on its configuration parameters
+	Hasher interface {
+		Hash(ctx context.Context) string
+	}
+
+	// A validater validates its own configuration parameters and returns an error if any inconsistencies are found
+	Validater interface {
+		Validate(ctx context.Context) error
+	}
+
+	// An initializer initializes internal data structures to prepare for runtime
+	Initializer interface {
+		Initialize(ctx context.Context) error
+	}
+
+	// A Filterer is a generic interface that can filter, match, transform, split or do any number of things with a given event
+	Filterer interface {
+		Filter(ctx context.Context, event *Event) ([]*Event, error) // the event slice can contain 0 (filter case), 1 (match or transform case) or n events (split case)
+	}
+
+	// An AsyncDoer processes things of similar type asynchronously, one after another, usually by listening on a blocking channel
+	AsyncDoer interface {
+		DoAsync(ctx context.Context)
+	}
+
+	// A SyncDoer processes things of similar type ynchronously, one after another, by explicit function call
+	SyncDoer interface {
+		DoSync(ctx context.Context, event *Event) error
+	}
+
+	// A matcher can match a pattern against an object
+	Matcher interface {
+		Match(ctx context.Context, event *Event, pattern interface{}) bool // if pattern is contained in event the function returns true
+	}
+
+	// An AckTree is a splittable acknowledge tree object
+	AckTree interface {
+		Ack()
+		SubTree(int) []AckTree
+	}
```

Comment:
please remove this (although I do like the idea of returning multiple SubTree 👍

Reply:
after merge (see above)